compiler/rustc_infer/src/infer/region_constraints/mod.rs

```diff
@@ -603,18 +628,21 @@ impl<'tcx> RegionConstraintCollector<'_, 'tcx> {
         tcx: TyCtxt<'tcx>,
         vid: ty::RegionVid,
     ) -> ty::Region<'tcx> {
-        let mut ut = self.unification_table_mut(); // FIXME(rust-lang/ena#42): unnecessary mut
+        let mut ut = self.unification_table_mut();
         let root_vid = ut.find(vid).vid;
-        let resolved = ut
-            .probe_value(root_vid)
-            .get_value_ignoring_universes()
-            .unwrap_or_else(|| ty::Region::new_var(tcx, root_vid));
-
-        // Don't resolve a variable to a region that it cannot name.
-        if self.var_universe(vid).can_name(self.universe(resolved)) {
```

Comment:
if we equate an infer var with a root which is in a higher universe, we previously did not return that root vid here. This caused the `PlaceholderReplacer` to not map a placeholder region back to a bound var in `deeply_normalize` due to the following setup:

- `for<'a> fn(RigidAlias<'a>)` gets instantiated as `fn(RigidAlias<'!a>)`
- normalizing `RigidAlias<'!a>` emits `AliasRelate(RigidAlias<'!a>, ?u)`
- uniquifying regions in the input results in the canonical goal `exists<U> exists<'a> AliasRelate(RigidAlias<'a>, U)`
- inside of the query we instantiate this to get `RigidAlias<'?a.1>` and `?u.0`
- `NormalizesTo` fails, so we first instantiate `?u.0` with `RigidAlias<'new.whatever>`, the universe of that gets pulled down into universe `0`
- we then equate the args of `RigidAlias<'new.0>` with `RigidAlias<'a.1>`, equating `'new` and `'a`, using `'a` as the root in the unification table
- when canonicalizing the response, we are unable to canonicalize `'new` as `'a` because the universe of `'a` is higher than the universe of `'new`
- because of this `RigidAlias<'!a>` normalizes to `RigidAlias<'new>` where `'new` is constrained to be equal to `'!a`
- we fail to map `'new` back to `'a` in the `PlaceholderReplacer`

Reply:
> * inside of the query we instantiate this to get `RigidAlias<'?a.1>` and `?u.0`

This looks like an issue with the new solver canonicalization! the universe of `?u` should certainly be able to name `'?a`. How does '?a` end up in U1 and `?u` as U0?

Reply:
In the new solver all regions are canonicalized to new inference variables in a single universe that is higher than everything else in the input. `?u` _can_ name `'?a` even with this setup it just requires lowering `'?a`'s universe when equating it which is something we already do for cases like `?x.0 eq ?y.1`. Even if the new solver did not have this canonicalization scheme I would still expect this PR to be required somehow as it is just wrong for `'?x.0 eq '?y.1` to not allow resolving `'?x` to `'?y` later on and there should be other ways to wind up with a region variable in a higher universe that is equated with something in a lower one